lib/app/server.ts

```diff
@@ -0,0 +1,437 @@
+import * as utils from '../utils.js';
+import type {
+    PugRequireHandler,
+    RenderConfig,
+    ServerDependencies,
+    ServerOptions,
+    WebServerResult,
+} from './server.interfaces.js';
+import {getFaviconFilename, isDevMode} from './utils.js';
+
+/**
+ * Detects if the request is from a mobile viewer
+ */
+export function isMobileViewer(req: Request): boolean {
+    return req.header('CloudFront-Is-Mobile-Viewer') === 'true';
+}
+
+/**
+ * Configure a web server and its routes
+ */
+export async function setupWebServer(
+    appArgs: AppArguments,
+    options: ServerOptions,
+    dependencies: ServerDependencies,
+): Promise<WebServerResult> {
+    const {ceProps, sponsorConfig, clientOptionsHandler} = dependencies;
+    const {
+        staticPath,
+        staticMaxAgeSecs,
+        staticUrl,
+        staticRoot,
+        httpRoot,
+        sentrySlowRequestMs,
+        distPath,
+        extraBodyClass,
+        maxUploadSize,
+    } = options;
+
+    // Create the Express application and router
+    const webServer = express();
+    const router = express.Router();
+
+    // Initialize services
+    const googleShortUrlResolver = new ShortLinkResolver();
+    let pugRequireHandler: PugRequireHandler = () => {
+        logger.error('pug require handler not configured');
+        return '';
+    };
+
+    /**
+     * Renders configuration for templates
+     */
+    function renderConfig(extra: Record<string, any>, urlOptions?: Record<string, any>): RenderConfig {
+        const urlOptionsAllowed = ['readOnly', 'hideEditorToolbars', 'language'];
+        const filteredUrlOptions = _.mapObject(_.pick(urlOptions || {}, urlOptionsAllowed), val =>
+            utils.toProperty(val),
+        );
+        const allExtraOptions = _.extend({}, filteredUrlOptions, extra);
+
+        if (allExtraOptions.mobileViewer && allExtraOptions.config) {
+            const clnormalizer = new normalizer.ClientStateNormalizer();
+            clnormalizer.fromGoldenLayout(allExtraOptions.config);
+            const clientstate = clnormalizer.normalized;
+
+            const glnormalizer = new normalizer.ClientStateGoldenifier();
+            allExtraOptions.slides = glnormalizer.generatePresentationModeMobileViewerSlides(clientstate);
+        }
+
+        const options = _.extend({}, allExtraOptions, clientOptionsHandler.get());
+        options.optionsHash = clientOptionsHandler.getHash();
+        options.compilerExplorerOptions = JSON.stringify(allExtraOptions);
+        options.extraBodyClass = options.embedded ? 'embedded' : extraBodyClass;
+        options.httpRoot = httpRoot;
+        options.staticRoot = staticRoot;
+        options.storageSolution = 'local'; // TODO: Don't hardcode this
+        options.require = pugRequireHandler;
+        options.sponsors = sponsorConfig;
+        return options;
+    }
+
+    /**
+     * Renders GoldenLayout for a given configuration
+     */
+    function renderGoldenLayout(
```

Comment:
Same with this one, createGoldenLayoutRenderer or something like that

Reply:
Likewise, we'd need to make a closure thing here. I'll add TODOs though

Reply:
Actually I tried to phrase the TODO and it was so complex (too many things called `options` in both cases)...

Reply:
I have made some inroads into improving this. The functions are still open though, mainly as we currently rely on the pughandler stuff (I think?)